File: Python/src/bqs/dijkstra_two_stack.py
from bqs.stack_linkedlist import StackLinkedList
from bqs.bqs_exception import BqsException


class DijkstraTwoStack:

    def __init__(self, infix):
        self.tokens = infix.split()
        self.opStack = StackLinkedList()
        self.valStack = StackLinkedList()
        self.parentheses = 0

    def process_token(self, s):
        if s == "(":
            self.parentheses += 1
        elif s == ")":
            self.parentheses -= 1
            self.operate()
        elif s in "+ - * / ^ %".split():
            self.opStack.push(s)
        else:
            n = int(s)
            self.valStack.push(n)

    def operate(self):
        y = self.valStack.pop()
        x = self.valStack.pop()

        if self.opStack.peek() == "+":
            self.valStack.push(x + y)
            self.opStack.pop()
            return
        elif self.opStack.peek() == "-":
            self.valStack.push(x - y)
            self.opStack.pop()
            return
        elif self.opStack.peek() == "*":
            self.valStack.push(x * y)
            self.opStack.pop()
            return
        elif self.opStack.peek() == "/":
            self.valStack.push(x / y)
            self.opStack.pop()
            return
        else:
            raise BqsException(f"operator not recognized: {self.opStack.peek()}")

    def evaluate(self):
        for i in self.tokens:
            self.process_token(i)

        if self.parentheses != 0:
            raise BqsException(f"there are {self.parentheses} superfluous parentheses (net)")

        while not self.opStack.is_empty():
            self.operate()

        result = self.valStack.pop()

        if not self.valStack.is_empty():
            raise BqsException("there are superfluous values")

        return result

File: src/main/java/edu/neu/coe/info6205/bqs/DijkstraTwoStack.java
package edu.neu.coe.info6205.bqs;

import java.util.StringTokenizer;

public class DijkstraTwoStack {

    private final StringTokenizer tokenizer;
    private int parentheses = 0;
    private final Stack<String> opStack = new Stack_LinkedList<>();
    private final Stack<Number> valStack = new Stack_LinkedList<>();

    public DijkstraTwoStack(String infix) {
        tokenizer = new StringTokenizer(infix);
    }

    public Number evaluate() throws BQSException {
        while (tokenizer.hasMoreTokens())
            processToken(tokenizer.nextToken());
        if (parentheses != 0)
            throw new BQSException("there are " + parentheses + " superfluous parentheses (net)");
        while (!opStack.isEmpty())
            operate();
        Number result = valStack.pop();
        if (!valStack.isEmpty())
            throw new BQSException("there are superfluous values");
        return result;
    }

    private void processToken(String s) throws BQSException {
        if (s.equals("("))
            parentheses++;
        else if (s.equals(")")) {
            parentheses--;
            operate();
        } else if ("+-*/^%".contains(s))
            opStack.push(s);
        else {
            Number n = Integer.parseInt(s);
            valStack.push(n);
        }
    }

    private void operate() throws BQSException {
        Integer y = (Integer) valStack.pop();
        Integer x = (Integer) valStack.pop();
        switch (opStack.pop()) {
            case "+":
                valStack.push(x + y);
                break;
            case "-":
                valStack.push(x - y);
                break;
            case "*":
                valStack.push(x * y);
                break;
            case "/":
                valStack.push(x / y);
                break;
            default:
                throw new BQSException("operator not recognized: " + opStack.pop());
        }
    }

    public static void main(String[] args) {
        try {
            DijkstraTwoStack twoStack = new DijkstraTwoStack("2 * ( 4 - 3 )");
            System.out.println(twoStack.evaluate());
        } catch (BQSException e) {
            e.printStackTrace();
        }
    }
}
